pilopt/tests/optimizer.rs

```diff
@@ -567,12 +634,159 @@ fn inline_chain_of_substitutions() {
+    // It CAN be inlined because high_degree has degree 1
+    // when calculating the degree of this expression
+    col witness depends_on_high;
+    depends_on_high = high_degree + x;
+    
+    // Usage of columns
+    x + a = 10;
+    high_degree + a = 20;
+    depends_on_high + a = 30;
+    depends_on_high + a = 31;
+"#;
+
+    // Expected result: x and depends_on_high can be inlined, but high_degree remains a witness
+    let expectation = r#"namespace N(65536);
+    col witness a;
+    col x = N::a * N::a;
+    col witness high_degree;
+    N::high_degree = N::a * N::a * N::a * N::a;
+    col depends_on_high = N::high_degree + N::x;
+    N::x + N::a = 10;
+    N::high_degree + N::a = 20;
+    N::depends_on_high + N::a = 30;
+    N::depends_on_high + N::a = 31;
+"#;
+
+    let optimized = optimize(analyze_string::<GoldilocksField>(input).unwrap(), 3).to_string();
+    assert_eq!(optimized, expectation);
+}
+
+#[test]
+fn multi_pass_optimization_unlocks_transformations() {
+    let input = r#"namespace N(65536);
+    col witness a;
+    
+    col witness early;
+    early = a * a;
+    
+    col witness middle;
+    middle = early * a;
+    
+    col witness valuable;
+    valuable = middle * middle;
+    
+    // Usage of columns
+    early + a = 10;
+    middle + a = 20;
+    valuable + a = 30;
+"#;
+
+    let expectation = r#"namespace N(65536);
+    col witness a;
+    col early = N::a * N::a;
+    col witness middle;
+    N::middle = N::early * N::a;
+    col valuable = N::middle * N::middle;
+    N::early + N::a = 10;
+    N::middle + N::a = 20;
+    N::valuable + N::a = 30;
+"#;
+
+    let optimized = optimize(analyze_string::<GoldilocksField>(input).unwrap(), 3).to_string();
+    assert_eq!(optimized, expectation);
+}
+
+#[test]
+fn avoid_removing_output_columns() {
+    let input = r#"
+        let N: int = 8;
+        
+        namespace Global(N); 
+            let alpha: expr = std::prelude::challenge(0, 41);
+            let beta: expr = std::prelude::challenge(0, 42);
+            col witness x;
+            col witness stage(1) y;
+            y = x * x + beta * alpha;
+        "#;
+
+    let expectation = r#"namespace Global(8);
+    col witness x;
+    col witness stage(1) y;
+    Global::y = Global::x * Global::x + std::prelude::challenge(0, 42) * std::prelude::challenge(0, 41);
+"#;
+
+    let optimized = optimize(analyze_string::<GoldilocksField>(input).unwrap(), 3).to_string();
+    assert_eq!(optimized, expectation);
+}
+
+#[test]
+fn preserve_input_output_witnesses() {
+    let input = r#"namespace N(65536);
+    col witness input_only;
+    col witness output_only;
+    col witness both_io;
+    col witness regular;
+    
+    // input_only only appears on the right side of equations
+    regular = input_only + 5;
+    output_only = input_only * 2;
+    
+    // output_only only appears on the left side of equations
+    output_only = regular + 3;
+    output_only = input_only * 2;
```

Comment:
Does this mean that the optimizer would perform differently if the sides are swapped?

Reply:
Wrong comments, fixed.